packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts

```diff
@@ -0,0 +1,719 @@
+  OpenId4VciCredentialSupportedWithId,
+  OpenId4VciIssuerMetadata,
+} from '../shared'
+import type {
+  AgentContext,
+  JwaSignatureAlgorithm,
+  W3cVerifiableCredential,
+  Key,
+  JwkJson,
+  SdJwtVc,
+} from '@aries-framework/core'
+import type {
+  AccessTokenResponse,
+  CredentialResponse,
+  Jwt,
+  OpenIDResponse,
+  PushedAuthorizationResponse,
+  AuthorizationDetails,
+  AuthorizationDetailsJwtVcJson,
+} from '@sphereon/oid4vci-common'
+
+import {
+  SdJwtVcApi,
+  getJwkFromJson,
+  DidsApi,
+  AriesFrameworkError,
+  Hasher,
+  InjectionSymbols,
+  JsonEncoder,
+  JwsService,
+  Logger,
+  SignatureSuiteRegistry,
+  TypedArrayEncoder,
+  W3cCredentialService,
+  W3cJsonLdVerifiableCredential,
+  W3cJwtVerifiableCredential,
+  getJwkClassFromJwaSignatureAlgorithm,
+  getJwkFromKey,
+  getKeyFromVerificationMethod,
+  getSupportedVerificationMethodTypesFromKeyType,
+  inject,
+  injectable,
+  parseDid,
+} from '@aries-framework/core'
+import {
+  AccessTokenClient,
+  CredentialRequestClientBuilder,
+  ProofOfPossessionBuilder,
+  formPost,
+  OpenID4VCIClient,
+} from '@sphereon/oid4vci-client'
+import { CodeChallengeMethod, ResponseType, convertJsonToURI, JsonURIMode } from '@sphereon/oid4vci-common'
+
+import { OpenId4VciCredentialFormatProfile } from '../shared'
+import {
+  getTypesFromCredentialSupported,
+  handleAuthorizationDetails,
+  getOfferedCredentials,
+} from '../shared/issuerMetadataUtils'
+import { getSupportedJwaSignatureAlgorithms } from '../shared/utils'
+
+import {
+  type OpenId4VciAuthCodeFlowOptions,
+  type OpenId4VciAcceptCredentialOfferOptions,
+  type OpenId4VciProofOfPossessionRequirements,
+  type OpenId4VciCredentialBindingResolver,
+  type OpenId4VciResolvedCredentialOffer,
+  type OpenId4VciResolvedAuthorizationRequest,
+  type OpenId4VciResolvedAuthorizationRequestWithCode,
+  type OpenId4VciSupportedCredentialFormats,
+  openId4VciSupportedCredentialFormats,
+} from './OpenId4VciHolderServiceOptions'
+
+@injectable()
+export class OpenId4VciHolderService {
+  private logger: Logger
+  private w3cCredentialService: W3cCredentialService
+  private jwsService: JwsService
+
+  public constructor(
+    @inject(InjectionSymbols.Logger) logger: Logger,
+    w3cCredentialService: W3cCredentialService,
+    jwsService: JwsService
+  ) {
+    this.w3cCredentialService = w3cCredentialService
+    this.jwsService = jwsService
+    this.logger = logger
+  }
+
+  public async resolveCredentialOffer(credentialOffer: string): Promise<OpenId4VciResolvedCredentialOffer> {
+    const client = await OpenID4VCIClient.fromURI({
+      uri: credentialOffer,
+      resolveOfferUri: true,
+      retrieveServerMetadata: true,
+    })
+
+    if (!client.credentialOffer?.credential_offer) {
+      throw new AriesFrameworkError(`Could not resolve credential offer from '${credentialOffer}'`)
+    }
+    const credentialOfferPayload: OpenId4VciCredentialOfferPayload = client.credentialOffer?.credential_offer
```

Comment:
```suggestion
    const credentialOfferPayload: OpenId4VciCredentialOfferPayload = client.credentialOffer.credential_offer
```

Assertion above should have confirmed it exists